File: 076 Minimum Window Substring.cpp
/*
Minimum Window Substring
https://leetcode.com/problems/minimum-window-substring/

Given a string S and a string T, find the minimum window in S which will contain all the characters in T in complexity O(n).

Example:

Input: S = "ADOBECODEBANC", T = "ABC"
Output: "BANC"
Note:

If there is no such window in S that covers all characters in T, return the empty string "".
If there is such window, you are guaranteed that there will always be only one unique minimum window in S.
*/

#include <iostream>
#include <string>
#include <limits.h>

typedef std::string STR;
typedef std::size_t INT;

STR minWin(STR src, STR tag){
    INT i,j;
    INT mask[256] = {0}; 
    INT found[256] = {0};
    INT minLen = INT_MAX;
    INT winlen = 0;
    INT count = 0;
    STR win = "";
    
    for(i =0;i<tag.size();i++) mask[tag[i]]++;
    i = 0;
    j = 0;
    for(;i<src.size();i++){
        if(mask[src[i]] > 0){
            found[src[i]]++;
            if(found[src[i]] <= mask[src[i]]) count++;
            
            //If found a candidate, shrink
            if(count == tag.size()){
                for(;mask[src[j]]==0 || found[src[j]] > mask[src[j]];j++)
                    if(found[src[j]] > mask[src[j]]) found[src[j]]--;
                
                //Update min win
                winlen = i-j+1;
                if(minLen > winlen){
                    minLen = winlen;
                    win = src.substr(j,winlen);
                }
            }
        }
    }
    return win;
}
    
int main(int argc, char** argv)
{
    STR S = "ADOBECODEBANC";
    STR T = "ABC";
    std::cout<<minWin(S,T)<<"\n";
    return 0;
}

File: 076 Minimum Window Substring.py
'''
76. Minimum Window Substring
https://leetcode.com/problems/minimum-window-substring/

Given two strings s and t of lengths m and n respectively, return the minimum window
substring of s such that every character in t (including duplicates) is included in the window.
If there is no such substring, return the empty string "".

The testcases will be generated such that the answer is unique.

Example 1:

Input: s = "ADOBECODEBANC", t = "ABC"
Output: "BANC"
Explanation: The minimum window substring "BANC" includes 'A', 'B', and 'C' from string t.

Example 2:

Input: s = "a", t = "a"
Output: "a"
Explanation: The entire string s is the minimum window.

Example 3:

Input: s = "a", t = "aa"
Output: ""
Explanation: Both 'a's from t must be included in the window.
Since the largest window of s only has one 'a', return empty string.

Constraints:
    m == s.length
    n == t.length
    1 <= m, n <= 105
    s and t consist of uppercase and lowercase English letters.
'''
class SlidingWindow:
    def __init__(self, pattern: str):
        self.patternCounter = Counter(pattern)
        self.patternSize = len(pattern)
        self.validCount = 0
        self.charCounts = Counter()
    
    def isMatched(self) -> bool:
        return self.validCount == self.patternSize
    
    def add(self, char: str):
        newCharCount = self.charCounts[char] + 1
        self.charCounts[char] = newCharCount
        if newCharCount <= self.patternCounter[char]:
            self.validCount += 1
    
    def remove(self, char: str):
        newCharCount = self.charCounts[char] - 1
        self.charCounts[char] = newCharCount
        if newCharCount < self.patternCounter[char]:
            self.validCount -= 1
    
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        window = SlidingWindow(t)
        result = ''
        left = 0
        for right, val in enumerate(s):
            # Extend
            window.add(val)

            while window.isMatched():
                # Result
                if result == '' or right - left + 1 < len(result):
                    result = s[left : right + 1]
                # Shrink
                window.remove(s[left])
                left += 1
        return result